src/OpenTelemetry.Contrib.Instrumentation.AWS/Implementation/AWSTracingPipelineHandler.cs

```diff
@@ -0,0 +1,270 @@
+        };
+
+        private static readonly Dictionary<string, string> ServiceParameterMap = new Dictionary<string, string>()
+        {
+            { "DynamoDBv2", "TableName" },
+            { "SQS", "QueueUrl" },
+        };
+
+        private static readonly Dictionary<string, string> ParameterAttributeMap = new Dictionary<string, string>()
+        {
+            { "TableName", AWSSemanticConventions.AttributeAWSDynamoTableName },
+            { "QueueUrl", AWSSemanticConventions.AttributeAWSSQSQueueUrl },
+        };
+
+        private static readonly ActivitySource AWSSDKActivitySource = new ActivitySource(ActivitySourceName);
+
+        private readonly AWSClientInstrumentationOptions options;
+
+        public AWSTracingPipelineHandler(AWSClientInstrumentationOptions options)
+        {
+            this.options = options;
+        }
+
+        public override void InvokeSync(IExecutionContext executionContext)
+        {
+            var activity = this.ProcessBeginRequest(executionContext);
+            try
+            {
+                base.InvokeSync(executionContext);
+            }
+            catch (Exception ex)
+            {
+                if (activity != null)
+                {
+                    this.ProcessException(activity, ex);
+                }
+
+                throw;
+            }
+            finally
+            {
+                if (activity != null)
+                {
+                    this.ProcessEndRequest(executionContext, activity);
+                }
+            }
+        }
+
+        public override async Task<T> InvokeAsync<T>(IExecutionContext executionContext)
+        {
+            T ret = null;
+
+            var activity = this.ProcessBeginRequest(executionContext);
+            try
+            {
+                ret = await base.InvokeAsync<T>(executionContext).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                if (activity != null)
+                {
+                    this.ProcessException(activity, ex);
+                }
+
+                throw;
+            }
+            finally
+            {
+                if (activity != null)
+                {
+                    this.ProcessEndRequest(executionContext, activity);
+                }
+            }
+
+            return ret;
+        }
+
+        private Activity ProcessBeginRequest(IExecutionContext executionContext)
+        {
+            Activity activity = null;
+
+            var requestContext = executionContext.RequestContext;
+            var service = this.GetAWSServiceName(requestContext);
+            var operation = this.GetAWSOperationName(requestContext);
+
+            activity = AWSSDKActivitySource.StartActivity(service + "." + operation, ActivityKind.Client);
+
+            if (activity == null)
+            {
+                return null;
+            }
+
+            if (this.options.SuppressDownstreamInstrumentation)
```

Comment:
what is the downsteam instrumentation in this case?

Reply:
AWS SDK requests are basically http requests underneath which is the downstream instrumentation here. If the customer appllication has http instrumentation enabled, not suppressing this downstream instrumentation will cause 2 spans to be created, one for AWS SDK instrumentation and one for underlying http request. So we are [defaulting to suppressing](https://github.com/srprash/opentelemetry-dotnet-contrib/blob/aws_sdk_instrumentation/src/OpenTelemetry.Contrib.Instrumentation.AWS/AWSClientInstrumentationOptions.cs#L25) the http instrumentation for the AWS request.